### PR TITLE
Support for AIO

### PR DESCRIPTION
**Pull request by khenderick**
This PR introduces support for a new module, code named AIO. It's basically a Gateway on steroids with build-in extras such as inputs, outputs, ...

First, code will be added next to existing code, often lacking backwards compatibility. This is mainly to speed up the development and testing. Next, code will be refactored in such way that it can run on both AIO and Gateway, dynamically figuring out the capabilities of the active hardware platform.

Note that code in this branch will most likely be unstable, might not even work at certain points and needs specific hardware in the early stages. It won't reflect the final feature set of this new AIO module, as features might be added for testing purposes, and be removed or changed along the way.

### CODE REVIEW

**Comment by khenderick** [comment]
At this point we made the decision to start integrating this branch into mainstream development. It introduces some cleaner code (paths) that will benefit the whole project. By merging it into develop as soon as possible, we'll gradually clean up code and make it work for both hardware platforms.

File: src/master/maintenance.py [hunk 1]
```diff
@@ -13,170 +13,130 @@
...
-
-    def reader(self):
+        logger.info('Activating maintenance mode')
+        self._last_maintenance_send_time = time.time()
+        self._master_communicator.start_maintenance_mode()
+        self._maintenance_timeout_timer = Timer(MaintenanceClassicCommunicator.MAINTENANCE_TIMEOUT, self._check_maintenance_timeout)
+        self._maintenance_timeout_timer.start()
+        self._stopped = False
+        self._read_data_thread = Thread(target=self._read_data, name='Classic maintenance read thread')
+        self._read_data_thread.daamon = True
```
Comment (yclybouw) [inline src/master/maintenance.py]: `daamon` → `daemon`

File: src/master_core/maintenance.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        self._deactivated_callback = None
+        self._maintenance_active = False
+        self._stopped = True
+        self._read_data_thread = None
+        self._active = False
+
+    def start(self):
+        self._stopped = False
+        self._read_data_thread = Thread(target=self._read_data, name='Core maintenance read thread')
+        self._read_data_thread.daamon = True
```
Comment (yclybouw) [inline src/master_core/maintenance.py]: `daamon` → `daemon`

File: src/gateway/observer.py [hunk 3]
```diff
@@ -154,15 +157,14 @@ def subscribe_events(self, callback):
...
+        if Platform.get_platform() == Platform.Type.CLASSIC:
+            self._thread.start()
 
     def invalidate_cache(self, object_type=None):
         """
         Triggered when an external service knows certain settings might be changed in the background.
         For example: maintenance mode or module discovery
         """
-        if object_type is None or object_type == Observer.Types.OUTPUTS:
-            self._output_last_updated = 0
```
[the remark this answers is not in the document]
Comment (khenderick) [reply src/gateway/observer.py]: Added this

File: src/master_core/ucan_updater.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import logging
+import os
+import struct
+import time
+from intelhex import IntelHex
+from master_core.ucan_api import UCANAPI
+from master_core.ucan_command import UCANPalletCommandSpec, SID
+from master_core.fields import Int32Field
+
+LOGGER = logging.getLogger('openmotics')
```
Comment (svanscho) [inline src/master_core/ucan_updater.py]: I typically use lower case logger names throughout the code. Can we standardize on 1 format?

File: src/master_core/memory_types.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+
+class MemoryVersionField(MemoryAddressField):
+    def __init__(self, memory_type, address_spec):
+        super(MemoryAddressField, self).__init__(memory_type, address_spec, 3)
+
+    def decode(self, data):
+        result = []
+        for item in data:
+            result.append('{0}'.format(item))
```
Comment (svanscho) [inline src/master_core/memory_types.py]: '.'.join(str(item) for item in data)

File: src/master_core/memory_types.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+            except ValueError:
+                raise ValueError(error_message)
+            if not (0 <= part <= 255):
+                raise ValueError(error_message)
+            data.append(part)
+        return data
+
+    def decode(self, data):
+        result = []
+        for item in data:
```
Comment (svanscho) [inline src/master_core/memory_types.py]: '.'.join('{0:03}'.format(item) for item in data)

File: src/master_core/errors.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        COMMAND_ERROR = 'COMMAND_ERROR'
+        UNKNOWN = 'UNKNOWN'
+
+    StateMachineTypes = [Types.SM_UPDATE_TIME_DATE, Types.SM_IMMEDIATE_QUEUE, Types.SM_GROUP_QUEUE, Types.SM_TIMER,
+                         Types.SM_EEPROM_ACTIVATE_STATE, Types.SM_PERFORM_EEPROM_ACTIVATE, Types.SM_CLI_PRINT,
+                         Types.SM_CAN_QUEUE, Types.SM_CAN, Types.SM_EXECUTE_EVERY_MINUTE, Types.SM_EXECUTE_GROUP_ACTION,
+                         Types.SM_GROUP_DELAY_QUEUE, Types.SM_CAN_TX_QUEUE]
+
+    def __init__(self, data):
+        self._parameter_a = data['parameter_a']
```
Comment (svanscho) [inline src/master_core/errors.py]: can we add a link to the wiki documentation for the # params?

File: src/master_core/core_communicator.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+        """ Get the number of seconds since the last successful communication. """
+        if self._last_success == 0:
+            return 0  # No communication - return 0 sec since last success
+        else:
+            return time.time() - self._last_success
+
+    def _get_cid(self):
+        """ Get a communication id. 0 and 1 are reserved. """
+        def _increment_cid(current_cid):
+            if current_cid is None:
```
Comment (svanscho) [inline src/master_core/core_communicator.py]: new_cid = current_cid + 1 if current_cid is not None and current_cid < 255) else 2

**Pull request merged**